cranelift/docs/isle-integration.md

```diff
@@ -0,0 +1,188 @@
+once we switch over to `regalloc2`, which takes instructions in SSA form
+directly as input.
+
+Instructions that implicitly operate on specific registers, or which require
+that certain operands be in certain registers, are handled similarly: the
+lowering rules use a pure paradigm that ignores these constraints and has
+instructions that explicitly take implicit operands, and we ensure the
+constraints are fulfilled a layer below the lowering rules (in `inst.isle` or in
+Rust glue code).
+
+## When are lowering rules allowed to have side effects?
+
+Extractors (the matchers that appear on the left-hand sides of `rule`s) should
+**never** have side effects. When evaluating a rule's extractors, we haven't yet
+committed to evaluating that rule's right-hand side. If the extractors performed
+side effects, we could get deeply confusing action-at-a-distance bugs where
+rules we never fully match pull the rug out from under our feet.
+
+Anytime you are tempted to perform side effects in an extractor, you should
+instead just package up the things you would need in order to perform that side
+effect, and then have a separate constructor that takes that package and
+performs the side effect it describes. The constructor can only be called inside
+a rule's right-hand side, which is only evaluated after we've committed to this
+rule, which avoids the action-at-a-distance bugs described earlier.
+
+For example, loads have a side effect in CLIF: they might trap. Therefore, even
+if a loaded value is never used, we will emit code that implements that
+load. But if we are compiling for x86 we can sink loads into other the operand
+for another operation depending on how the loaded value is used. If we sink that
+load into, say, an `add` then we need to tell the lowering context *not* to
+lower the CLIF `load` instruction anymore, because its effectively already
+lowered as part of lowering the `add` that uses the loaded value. That is a side
+effect, and we might be tempted to perform that side effect in the extractor
+that matches sinkable loads. But we can't do that because although the load
+itself might be sinkable, there might be a reason why we ultimately don't
+perform this load-sinking rule, and if that happens we still need to lower the
+CLIF load. So we make the `sinkable_load` extractor create a `SinkableLoad` type
+that packages up everything we need to know about the load and how to tell the
+lowering context that we've sunk it and the lowering context doesn't need to
+lower it anymore, but *it doesn't actually tell that to the lowering context
+yet*. Then, we pair that with a `sink_load` constructor that takes the
+`SinkableLoad`, performs the associated side effect of telling the lowering
+context not to lower the load anymore, and returns the x86 operand with the load
+sunken into it. See `sinkable_load`, `SinkableLoad`, and `sink_load` inside
+`cranelift/codegen/src/isa/x64/inst.isle` for details.
```

Comment:
This `SinkableLoad` discussion could refer to the section below as well, noting that it's a form of type safety, disallowing use of the value until we commit to merging it :-)